### ISSUE TITLE
 LightLLM  maye should provide a script to download the weight and convert the weight.

### ISSUE THREAD

**Issue by lambda7xx** [ModelTC/lightllm#37]
I found there is no script that can download the weight and convert the weight.

**Comment by lambda7xx** [comment]
Currently I try this [test_llama](https://github.com/ModelTC/lightllm/blob/main/test/model/test_llama.py) to run a test, but I am confused about the model weight. Use hugging face to download the weights? But [load_hf_weight](https://github.com/ModelTC/lightllm/blob/main/lightllm/models/llama/layer_weights/transformer_layer_weight.py#L26) seems we needs covert the huggingface model weight.

**Comment by llehtahw** [comment]
@lambda7xx  have you encountered any failure running the test with weights downloaded from hugging face?

**Comment by lambda7xx** [comment]
I use the follow code to download a llama-7b-half model. Currently its weight can not use for lightllm. we have to convert it

```

from transformers import AutoModelForCausalLM
model = AutoModelForCausalLM.from_pretrained("decapoda-research/llama-7b-hf")
```

**Comment by lambda7xx** [comment]
my  error.

```
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ubuntu/lightllm/test/model/model_infer.py", line 28, in tppart_model_infer
    torch.cuda.set_device(rank_id)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/torch/cuda/__init__.py", line 326, in set_device
    torch._C._cuda_setDevice(device)
RuntimeError: CUDA error: invalid device ordinal
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ubuntu/lightllm/test/model/model_infer.py", line 28, in tppart_model_infer
    torch.cuda.set_device(rank_id)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/torch/cuda/__init__.py", line 326, in set_device
    torch._C._cuda_setDevice(device)
RuntimeError: CUDA error: invalid device ordinal
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.

```

**Comment by llehtahw** [comment]
Could you please post your `nvidia-smi` output?

I guess your number of graphics cards is less than 8. You may set a smaller `world_size` value here.

https://github.com/ModelTC/lightllm/blob/3d84a6e625dbac541fb38957ec14726cb680ac84/test/model/test_llama.py#L11

@lambda7xx

**Comment by lambda7xx** [comment]
my GPUs is A10. I found the weight format from huggingface is wrong.

```
ss 'lightllm.models.llama.layer_infer.model.LlamaTpPartModel'>, batch_size:20, input_len:64, output_len:16
Process Process-1:
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ubuntu/lightllm/test/model/model_infer.py", line 34, in tppart_model_infer
    model_part = model_class(dist.get_rank(),
  File "/home/ubuntu/.local/lib/python3.10/site-packages/lightllm-1.0.0-py3.10.egg/lightllm/models/llama/layer_infer/model.py", line 20, in __init__
    with open(os.path.join(weight_dir, "config.json"), 'r') as json_file:
FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/.cache/huggingface/hub/models--decapoda-research--llama-7b-hf/blobs/config.json'
F
```

**Comment by lambda7xx** [comment]
if we download the weight from  https://huggingface.co/decapoda-research/llama-7b-hf/tree/main, its format is bin

**Comment by llehtahw** [comment]
> if we download the weight from https://huggingface.co/decapoda-research/llama-7b-hf/tree/main, its format is bin

https://github.com/ModelTC/lightllm/blob/3d84a6e625dbac541fb38957ec14726cb680ac84/lightllm/models/llama/layer_weights/hf_load_utils.py#L15-L18

bin format is also supported.

> FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/.cache/huggingface/hub/models--decapoda-research--llama-7b-hf/blobs/config.json'

This indicates the model path you passed in might be wrong.

What `model_dir` value did you set here?

https://github.com/ModelTC/lightllm/blob/3d84a6e625dbac541fb38957ec14726cb680ac84/test/model/test_llama.py#L12

And you may run `find /home/ubuntu/.cache/huggingface/hub/models--decapoda-research--llama-7b-hf -name config.json` to check if you have set the right `model_dir` value.

**Comment by lambda7xx** [comment]
let me try again.

**Comment by HalFTeen** [comment]
@lambda7xx I use the same way with you to download the llma-7b-hf, however tokenizer is not download successfully, as your error log mentioned: can't load tokenizer. you should check if the tokenizer.model is existed or not.

**Comment by llehtahw** [comment]
> @lambda7xx I use the same way with you to download the llma-7b-hf, however tokenizer is not download successfully, as your error log mentioned: can't load tokenizer. you should check if the tokenizer.model is existed or not.

And you can also download the whole repo with git @lambda7xx 

```sh
# Make sure you have git-lfs installed (https://git-lfs.com)
git lfs install
git clone https://huggingface.co/decapoda-research/llama-7b-hf

# if you want to clone without large files – just their pointers
# prepend your git clone with the following env var:
GIT_LFS_SKIP_SMUDGE=1
```

**Comment by lambda7xx** [comment]
@llehtahw  I use your command to download the weights.
I still have this problem

```
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/ubuntu/lightllm/lightllm/server/api_server.py", line 196, in <module>
    main()
  File "/home/ubuntu/lightllm/lightllm/server/api_server.py", line 163, in main
    httpserver_manager = HttpServerManager(args.model_dir,
  File "/home/ubuntu/lightllm/lightllm/server/httpserver/manager.py", line 18, in __init__
    self.tokenizer = get_tokenizer(model_weightdir, tokenizor_mode)
  File "/home/ubuntu/lightllm/lightllm/server/tokenizer.py", line 55, in get_tokenizer
    tokenizer = AutoTokenizer.from_pretrained(tokenizer_name, *args,
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/auto/tokenization_auto.py", line 690, in from_pretrained
    raise ValueError(
ValueError: Tokenizer class LLaMATokenizer does not exist or is not currently imported.
```

**Comment by lambda7xx** [comment]
- I use your patch, then I use the command

```
python3 -m lightllm.server.api_server --model_dir /home/ubuntu/llama-7b-hf --tp 1 --max_total_token_num 120000

```
- my error log 
```
Using a slow tokenizer. This might cause a significant slowdown. Consider using a fast tokenizer instead.
init MemoryManager size 120000 dtype torch.float16 head_num 32 head_dim 128 layer_num 32
init MemoryManager, self.mem_state.shape:torch.Size([120000]) and self._mem_cum_sum.shape:torch.Size([120000]) and self.indexs.shape:torch.Size([120000]) can_use_mem_size 120000
Process Process-1:
router init state: CUDA out of memory. Tried to allocate 938.00 MiB (GPU 0; 21.99 GiB total capacity; 21.07 GiB already allocated; 601.00 MiB free; 21.07 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF detoken init state: init ok
```
I change the 120000 to 12000, still fails.
```
Using a slow tokenizer. This might cause a significant slowdown. Consider using a fast tokenizer instead.
Using a slow tokenizer. This might cause a significant slowdown. Consider using a fast tokenizer instead.
init MemoryManager size 12000 dtype torch.float16 head_num 32 head_dim 128 layer_num 32
init MemoryManager, self.mem_state.shape:torch.Size([12000]) and self._mem_cum_sum.shape:torch.Size([12000]) and self.indexs.shape:torch.Size([12000]) can_use_mem_size 12000
router init state: 'list' object has no attribute 'shape' detoken init state: init ok
```
It means I can't use one A10 24GPUs to server a 7b model. @llehtahw

**Comment by llehtahw** [comment]
> router init state: 'list' object has no attribute 'shape' detoken init state: init ok

@hiworldwzj please have a look

**Comment by hiworldwzj** [comment]
> * I use your patch, then I use the command
> 
> ```
> python3 -m lightllm.server.api_server --model_dir /home/ubuntu/llama-7b-hf --tp 1 --max_total_token_num 120000
> ```
> 
> * my error log
> 
> ```
> Using a slow tokenizer. This might cause a significant slowdown. Consider using a fast tokenizer instead.
> init MemoryManager size 120000 dtype torch.float16 head_num 32 head_dim 128 layer_num 32
> init MemoryManager, self.mem_state.shape:torch.Size([120000]) and self._mem_cum_sum.shape:torch.Size([120000]) and self.indexs.shape:torch.Size([120000]) can_use_mem_size 120000
> Process Process-1:
> router init state: CUDA out of memory. Tried to allocate 938.00 MiB (GPU 0; 21.99 GiB total capacity; 21.07 GiB already allocated; 601.00 MiB free; 21.07 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF detoken init state: init ok
> ```
> 
> I change the 120000 to 12000, still fails.
> 
> ```
> Using a slow tokenizer. This might cause a significant slowdown. Consider using a fast tokenizer instead.
> Using a slow tokenizer. This might cause a significant slowdown. Consider using a fast tokenizer instead.
> init MemoryManager size 12000 dtype torch.float16 head_num 32 head_dim 128 layer_num 32
> init MemoryManager, self.mem_state.shape:torch.Size([12000]) and self._mem_cum_sum.shape:torch.Size([12000]) and self.indexs.shape:torch.Size([12000]) can_use_mem_size 12000
> router init state: 'list' object has no attribute 'shape' detoken init state: init ok
> ```
> 
> It seems I can't use one A10 24GPUs to server a 7b model. @llehtahw

max_total_token_num should be more smaller。please see https://github.com/ModelTC/lightllm/blob/main/docs/ApiServerArgs.md
